src/shared/watchdog.c

```diff
@@ -133,6 +133,8 @@ static int open_watchdog(void) {
         fn = !watchdog_device || path_equal(watchdog_device, "/dev/watchdog") ?
                 "/dev/watchdog0" : watchdog_device;
 
+        watchdog_set_device(fn);
```

Comment:
if `fn` is `watchdog_device` anway we'll strdup() it and free() the old identical string. Also the call can fail due to OOM. Maybe just do this instead:

```c
if (!streq_ptr(fn, watchdog_device)) {
        /* Keep track of the device we are actually opening */ 
        r = free_and_strdup(&watchdog_device, fn);
        if (r < 0)
                return log_debug_debug();
}
```

Reply:
`free_and_strdup()` already does a `streq_ptr()` check internally. If the string matches, then it does nothing.

I changed it to call `free_and_strdup()` anyways and added the oom check.

Reply:
then I'm not sure to understand why you didn't stick with `watchdog_set_device(fn);` instead... (with the return value check thought).

Or maybe make it possibly slightly clearer by calling `watchdog_set_device()` only when fn is "dev/watchdog0".

Reply:
then I'm not sure to understand why you didn't stick with watchdog_set_device(fn); instead... (with the return value check thought).

you could, but the closing of the fd isn't really necessary then. With @michaelolbrich' current version you have exactly one function call with error handling. if this would use watchdog_set_device() instead you'd have the exact same number of function calls, and it would do stuff we don't really need.